### PR TITLE
don't support multiple DMA regions per map/unmap

### PR DESCRIPTION
**Pull request by tmakatos**
Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

**Comment by tmakatos** [comment]
Need to restore `argsz` and to ensure compatibility with QEMU.

File: lib/libvfio-user.c [hunk 1]
```diff
@@ -476,90 +476,102 @@ consume_fd(int *fds, size_t nr_fds, size_t index)
...
-handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+handle_dma_map(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
+               struct vfio_user_dma_region *region)
 {
-    struct vfio_user_dma_region *dma_regions = msg->in_data;
-    bool is_map = (msg->hdr.cmd == VFIO_USER_DMA_MAP);
-    size_t nr_dma_regions;
-    size_t fdi;
-    size_t i;
+    char rstr[1024]; /* TODO this can be much smaller */;
```
Comment (jlevon) [inline lib/libvfio-user.c:482]: nit, please either dial down, or just remove the TODO, it's not really a big deal.

File: lib/libvfio-user.c [hunk 2]
```diff
@@ -476,90 +476,102 @@ consume_fd(int *fds, size_t nr_fds, size_t index)
...
+            close(fd);
+        }
+        return ERROR_INT(ret);
     }
 
-    for (i = 0, fdi = 0; i < nr_dma_regions; i++) {
-        struct vfio_user_dma_region *region = &dma_regions[i];
-        char rstr[1024];
+    if (vfu_ctx->dma_register != NULL) {
+        vfu_ctx->dma_register(vfu_ctx,
```
Comment (jlevon) [inline lib/libvfio-user.c:526]: nit, does this fit on one line now?

File: lib/libvfio-user.c [hunk 3]
```diff
@@ -476,90 +476,102 @@ consume_fd(int *fds, size_t nr_fds, size_t index)
...
-            if (ret < 0) {
-                ret = errno;
-                vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region %s: %m",
-                        rstr);
-                if (fd != -1) {
-                    close(fd);
-                }
-                return ERROR_INT(ret);
-            }
+    char rstr[1024];
```
Comment (jlevon) [inline lib/libvfio-user.c:548]: nit, I don't think we've been using C99-style here, should be at the top of the func

File: lib/libvfio-user.c [hunk 4]
```diff
@@ -476,90 +476,102 @@ consume_fd(int *fds, size_t nr_fds, size_t index)
...
-                                               vfu_ctx->dma_unregister,
-                                               vfu_ctx);
-            if (ret < 0) {
-                ret = errno;
-                vfu_log(vfu_ctx, LOG_ERR, "failed to remove DMA region %s: %m",
-                        rstr);
-                return ERROR_INT(ret);
-            }
+    if (region->flags == 0) {
+        if (msg->in_size != sizeof(struct vfio_user_dma_region)) {
```
Comment (jlevon) [inline lib/libvfio-user.c:557]: you already checked this at :542

**Review by jlevon — CHANGES_REQUESTED** [review]
just a couple of nits

File: lib/libvfio-user.c (hunk 1 above)
Comment (tmakatos) [reply lib/libvfio-user.c:482]: Sure, I'll just remove the TODO.

File: lib/libvfio-user.c (hunk 2 above)
Comment (tmakatos) [reply lib/libvfio-user.c:526]: yes

File: lib/libvfio-user.c (hunk 3 above)
Comment (tmakatos) [reply lib/libvfio-user.c:548]: right

File: lib/libvfio-user.c (hunk 4 above)
Comment (tmakatos) [reply lib/libvfio-user.c:557]: you're right

**Comment by tmakatos** [comment]
@jlevon I've addressed your comments, please take another look.

**Review by jlevon — APPROVED** [review]
Let's get this in, even if we need to iterate over the exact format still. Thanks!

**Pull request merged**